### DISCUSSION TITLE
Visualizing/evaluating output of centroid model

### DISCUSSION THREAD

**Discussion by deligkarisk** [talmolab/sleap#1364] [Help!] [answered]
Hi,

I am looking to re-create Figure 3b (stage 2) figure of the SLEAP publication with my own data. Briefly, I would like to visualize the output of the centroid model in order to evaluate whether my configuration is appropriate. I am therefore looking to extract the location of the centroid/anchor point, as well as the coordinates of the bounding box that was used to crop the image before being sent to the centered-instance model.

One approach I tried was to load both the centroid model and labels project, do  the predictions, and loop through the resulting labeled_frames's predicted_instances. The later contain "bounding_box" and "centroid" properties, which I originally thought are what I need to look at. However, I am somewhat unclear if this is indeed the case, as I am not sure how to interpret the concept of predicted instances as a result of the centroid model only output.

I would appreciate if someone could point me in the right direction for visualizing the output of the centroid model  (i.e. extracting the centroid/anchor point coordinates and bounding/crop box).

**Comment by talmo** [comment, marked as answer]
Hi @deligkarisk,

Not a full answer to your question, but just to get you started, [here's a Colab notebook](https://colab.research.google.com/drive/1293_lXPXoHzP1sPS6tizIhr-so8bduUW) that shows how to get the low level outputs from the models for a bottom-up model. You can just modify it for a top-down model and be able to get the confidence maps for the centroids out as well.

You'll be using the [`TopDownInferenceModel`](https://sleap.ai/api/sleap.nn.inference.html#sleap.nn.inference.TopDownInferenceModel) class (accessible via `predictor.inference_model`) and its constituent layers [`CentroidCrop`](https://sleap.ai/api/sleap.nn.inference.html#sleap.nn.inference.CentroidCrop) and [`FindInstancePeaks`](https://sleap.ai/api/sleap.nn.inference.html#sleap.nn.inference.FindInstancePeaks) which have attributes that allow you to configure what's being returned when you run prediction, including the intermediate outputs you need.

Give this a spin based on that Colab and the info above and feel free to let us know if you have any questions!

**Comment by deligkarisk** [comment]
Hi @talmo, 

Thank you very much for your reply, I was able to get started and got a feeling of what I need to do.
I do have a follow-up question on figuring out the crop size used so that I can recreate what happened during inference through the command-line.

At the moment I am just loading the centroid model by itself. In the constructor of the TopDownPredictor class, if the centered instance model is not present, then the crop_size is set as equal to one. Which leads to the crops just being one pixel.

What I am thinking of doing therefore is to load the centered model, get the crop size from the configuration data and then set this to the centroid model crop size. (I have set the crop size to auto in the configuration files). Does that sound more or less a good way to go?

**Reply by talmo** [reply]
Yep, if it helps you can use [`sleap.load_config`](https://sleap.ai/api/sleap.nn.config.training_job.html#sleap.nn.config.training_job.load_config) to load just the configuration without the whole model if you need to query for the crop size.

**Reply by deligkarisk** [reply]
OK, got it. Thank you very much for your feedback.
